File: exam-slammr-webapp/src/main/java/uk/co/claritysoftware/exam/slammr/webapp/web/model/question/CreateQuestion.java
package uk.co.claritysoftware.exam.slammr.webapp.web.model.question;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption.generateEmptyAnswerOption;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading.generateEmptyFurtherReading;

import java.util.List;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Class encapsulating fields within a request to create a new question
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public final class CreateQuestion {

	@NotNull
	private Action action;

	@NotEmpty(message = "{CreateQuestion.summary.NotEmpty.message}")
	private String summary;

	@NotEmpty(message = "{CreateQuestion.question.NotEmpty.message}")
	private String question;

	@Size(min = 2)
	private List<AnswerOption> answerOptions;

	private List<String> tags;

	private List<String> certifications;

	private List<FurtherReading> furtherReadings;

	/**
	 * @return a new CreateQuestion instance with empty/default fields
	 */
	public static CreateQuestion generateEmptyCreateQuestion() {
		return CreateQuestion.builder()
				.answerOptions(asList(
						generateEmptyAnswerOption(),
						generateEmptyAnswerOption()
				))
				.tags(singletonList(""))
				.certifications(singletonList(""))
                .furtherReadings(singletonList(generateEmptyFurtherReading()))
				.build();
	}

	public enum Action {
		save, addTag, addCertification, addFurtherReading, addAnswer
	}
}

File: exam-slammr-webapp/src/main/kotlin/uk/co/claritysoftware/exam/slammr/webapp/web/model/question/CreateQuestion.kt
package uk.co.claritysoftware.exam.slammr.webapp.web.model.question

import javax.validation.constraints.NotBlank
import javax.validation.constraints.Size

/**
 * Class encapsulating fields within a request to create a new question
 */
data class CreateQuestion(var action: Action? = null,

						  @field:NotBlank(message = "{CreateQuestion.summary.NotBlank.message}")
						  var summary: String = "",

						  @field:NotBlank(message = "{CreateQuestion.question.NotBlank.message}")
						  var question: String = "",

						  @field:Size(min = 2)
						  var answerOptions: MutableList<AnswerOption> = mutableListOf(AnswerOption()),

						  var tags: MutableList<String> = mutableListOf(""),

						  var certifications: MutableList<String> = mutableListOf(""),

						  var furtherReadings: MutableList<FurtherReading> = mutableListOf(FurtherReading())
) {

	enum class Action {
		save, addTag, addCertification, addFurtherReading, addAnswer
	}
}